datafusion/src/row/reader.rs

```diff
@@ -0,0 +1,441 @@
+                let offset = self.field_offsets[idx];
+                let start = self.base_offset + offset;
+                let end = start + $WIDTH;
+                $NATIVE::from_le_bytes(self.data[start..end].try_into().unwrap())
+            }
+        }
+    };
+}
+
+macro_rules! fn_get_idx_opt {
+    ($NATIVE: ident) => {
+        paste::item! {
+            fn [<get_ $NATIVE _opt>](&self, idx: usize) -> Option<$NATIVE> {
+                if self.is_valid_at(idx) {
+                    Some(self.[<get_ $NATIVE>](idx))
+                } else {
+                    None
+                }
+            }
+        }
+    };
+}
+
+struct RowReader<'a> {
+    data: &'a [u8],
+    base_offset: usize,
+    field_count: usize,
+    null_width: usize,
+    field_offsets: Vec<usize>,
+}
+
+impl<'a> std::fmt::Debug for RowReader<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let null_bits = self.null_bits();
+        super::bitmap::fmt(null_bits, 0, self.null_width, f)
+    }
+}
+
+impl<'a> RowReader<'a> {
+    fn new(schema: &Arc<Schema>, data: &'a [u8]) -> Self {
+        assert!(supported(schema));
+        let field_count = schema.fields().len();
+        let null_width = bytes_for(field_count);
+        let (field_offsets, _) = get_offsets(null_width, schema);
+        Self {
+            data,
+            base_offset: 0,
+            field_count,
+            null_width,
+            field_offsets,
+        }
+    }
+
+    /// Update this row to point to position `offset` in `base`
+    fn point_to(&mut self, offset: usize) {
+        self.base_offset = offset;
+    }
+
+    #[inline]
+    fn assert_index_valid(&self, idx: usize) {
+        assert!(idx < self.field_count);
+    }
+
+    #[inline(always)]
+    fn null_bits(&self) -> &[u8] {
+        let start = self.base_offset;
+        &self.data[start..start + self.null_width]
+    }
+
+    #[inline(always)]
+    fn all_valid(&self) -> bool {
+        let null_bits = self.null_bits();
+        all_valid(null_bits, self.field_count)
+    }
+
+    fn is_valid_at(&self, idx: usize) -> bool {
+        unsafe { get_bit_unchecked(self.null_bits(), idx) }
+    }
+
+    fn get_bool(&self, idx: usize) -> bool {
+        self.assert_index_valid(idx);
+        let offset = self.field_offsets[idx];
+        let value = &self.data[self.base_offset + offset..];
+        value[0] != 0
+    }
+
+    fn get_u8(&self, idx: usize) -> u8 {
+        self.assert_index_valid(idx);
+        let offset = self.field_offsets[idx];
+        self.data[self.base_offset + offset]
+    }
+
+    fn_get_idx!(u16, 2);
+    fn_get_idx!(u32, 4);
+    fn_get_idx!(u64, 8);
+    fn_get_idx!(i8, 1);
+    fn_get_idx!(i16, 2);
+    fn_get_idx!(i32, 4);
+    fn_get_idx!(i64, 8);
+    fn_get_idx!(f32, 4);
+    fn_get_idx!(f64, 8);
+
+    fn get_date32(&self, idx: usize) -> i32 {
+        get_idx!(i32, self, idx, 4)
+    }
+
+    fn get_date64(&self, idx: usize) -> i64 {
```

Comment:
Perhaps these methods should assert that `idx` is actually the corresponding type. Otherwise this is effectively reinterpreting memory as different primitives, which whilst technically safe, is borderline unsafe :grinning: